### ISSUE TITLE
Allow train_loop_config to be a dataclass / pydantic model

### ISSUE THREAD

**Issue by daturkel** [ray-project/ray#39824]
(edit: I've updated this issue from suggesting only using a pydantic model to allowing either pydantic or a python dataclass)

### Description

Right now, all arguments passed to a Ray Train `train_loop_per_worker` are meant to be `dict[str,Any]`. However, this type means that developer tools have no idea what keys to expect in the config and what their types are.

It would be nice to be able to (optionally) define the `train_loop_config` as a Pydantic model or native python dataclass:

```python
from pydantic import BaseModel

class TrainConfig(BaseModel):
    lr: float
    layer_sizes: list[int]
    loss: Literal["bce", "mse"]

train_config = TrainConfig(lr=0.2, layer_sizes=[32,32,16], loss="bce")

def training_func(config: TrainConfig):
    ... # now my editor knows exactly what keys and types are in `config`

[...]

trainer = TorchTrainer(
    train_loop_per_worker=my_func,
    train_loop_config=train_config
)
```

Note: Pydantic is [already supported](https://docs.ray.io/en/latest/serve/advanced-guides/app-builder-guide.html#typing-arguments-with-pydantic) as a way of passing arguments in Ray Serve.

### Use case

The use case for this feature is to enable better developer ergonomics: currently you have to manually make note of what keys are in the config passed to the train function, but with a Pydantic model or dataclass you would know because the class has a known structure.

An added benefit is to use [field metadata](https://stackoverflow.com/a/67112188) to indicate that some arguments should not be logged when Ray logs the input to the training function (e.g. if you are passing a secret to the training function).

**Comment by asoans** [comment]
Did you want to use dataclasses instead of pydantic models? You can try 
`
from dataclasses import dataclass

class TrainConfig;
lr: float
layer_sizes: list[int]
loss: str

train_config = TrainConfig(lr=0.2, layer_sizes=[32, 32, 16], loss="bce")
 `
and then pass the dataclass into the TorchTrainer

**Comment by daturkel** [comment]
@asoans Yes it's true, I can pass a dataclass currently to the `train_loop_config` of `TorchTrainer`, but there are two minor flaws currently:

1. `train_loop_config` is annotated (and documented) as `Optional[Dict]`, which means that type checking code will note this as an error https://github.com/ray-project/ray/blob/master/python/ray/train/torch/torch_trainer.py#L179
2. Ray's built in logging will not properly log the individual keys and values inside the dataclass:
```
Training started with configuration:
╭──────────────────────────────────────────╮
│ Training config                          │
├──────────────────────────────────────────┤
│ train_loop_config   ...sparse': True}}}} │
╰──────────────────────────────────────────╯
```

So to get dataclasses supported as a "first class citizen", the [underlying code to print this table](https://github.com/krfricke/ray/blob/master/python/ray/tune/experimental/output.py#L566) would need to be tweaked to support a dataclass, not just a dict.

**Comment by asoans** [comment]
That makes sense. Would it be possible if I could have a go at it?

**Comment by daturkel** [comment]
> That makes sense. Would it be possible if I could have a go at it?
Absolutely, thanks!

**Comment by matthewdeng** [comment]
This makes a lot of sense. One reason we haven't gotten around to doing this is that there's a bit of a complexity under the hood mapping that converts between these configurations, Ray Tune-able parameters, and results. For the time being, I would recommend converting between dataclass/pydantic model and dictionary outside of the Ray Train code.

**Comment by daturkel** [comment]
> This makes a lot of sense. One reason we haven't gotten around to doing this is that there's a bit of a complexity under the hood mapping that converts between these configurations, Ray Tune-able parameters, and results. For the time being, I would recommend converting between dataclass/pydantic model and dictionary outside of the Ray Train code.

That makes sense to me. As I'm currently using Train without Tune, I've actually just been abusing the current system by passing my dataclass as a value in the train config dictionary. It doesn't serialize correctly in the results, but it makes its way to the trainer intact. However, once we start doing tuning jobs, I'll likely need to do as you say and start passing real dicts. It just becomes unwieldy once that dict gets really big (and one advantage of the dataclass is that it can have methods which help select exactly the args i need for individual functions inside the training loop). I'll have to think about this a little more.

**Comment by cszhu** [comment]
Hello! This P3 issue has seen no activity in at least a year. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
